code/midtown/mmdyna/poly.cpp

```diff
@@ -19,3 +19,193 @@
+{
+    f32 max_y = -999999.0f;
+    i32 vert_count = (Flags & 4) != 0 ? 4 : 3;
+
+    for (i32 i = 0; i < vert_count; ++i)
+    {
+        const Vector3& cur_vert = mmBoundTemplate::VertPtr[VertIndices[i % vert_count]];
+        const Vector3& next_vert = mmBoundTemplate::VertPtr[VertIndices[(i + 1) % vert_count]];
+
+        if ((cur_vert.z < z_plane && next_vert.z > z_plane) || (cur_vert.z > z_plane && next_vert.z < z_plane))
+        {
+            f32 factor = (z_plane - cur_vert.z) / (next_vert.z - cur_vert.z);
+            f32 x_int = (next_vert.x - cur_vert.x) * factor + cur_vert.x;
+
+            if (x_int >= x_min && x_int <= x_max)
+            {
+                f32 y_int = (next_vert.y - cur_vert.y) * factor + cur_vert.y;
+
+                if (y_int > max_y)
+                {
+                    max_y = y_int;
+                }
+            }
+        }
+    }
+
+    return max_y;
+}
+
+f32 mmPolygon::CheckCorner(f32 x, f32 z, f32* x_plane, f32* z_plane, f32* const_plane)
+{
+    if (PlaneN.y == 0.0f)
+    {
+        return -999999.0f;
+    }
+
+    i32 vert_count = (Flags & 4) ? 4 : 3;
+
+    for (i32 i = 0; i < vert_count; ++i)
+    {
+        if (x * x_plane[i] + z * z_plane[i] + const_plane[i] < 0.0f)
+        {
+            return -999999.0f;
+        }
+    }
+
+    return -(PlaneN.x * x + PlaneD + PlaneN.z * z) / PlaneN.y;
+}
+
+f32 mmPolygon::CornersHeight(f32 x1, f32 z1, f32 x2, f32 z2)
+{
+    f32 max_y = -999999.0f;
+    i32 vert_count = (Flags & 4) ? 4 : 3;
+    f32 adjusted_sign = (PlaneN.y <= 0.0f) ? 1.0f : -1.0f;
+    f32 x_plane[4], z_plane[4], const_plane[4], length, inverse_length;
+
+    for (i32 i = 0; i < vert_count; ++i)
+    {
+        Vector3& cur_vert = mmBoundTemplate::VertPtr[VertIndices[i]];
+        Vector3& next_vert = mmBoundTemplate::VertPtr[VertIndices[(i + 1) % vert_count]];
+
+        x_plane[i] = -(next_vert.z - cur_vert.z) * adjusted_sign;
+        z_plane[i] = (next_vert.x - cur_vert.x) * adjusted_sign;
+        const_plane[i] = -(x_plane[i] * cur_vert.x + z_plane[i] * cur_vert.z);
+
+        length = std::sqrt(x_plane[i] * x_plane[i] + z_plane[i] * z_plane[i]);
+
+        if (length < 1e-9)
+            return -999999.0f;
+
+        inverse_length = 1.0f / length;
+        x_plane[i] *= inverse_length;
+        z_plane[i] *= inverse_length;
+        const_plane[i] *= inverse_length;
+    }
+
+    for (i32 i = 0; i < 4; ++i)
+    {
+        f32 corner_height;
+
+        corner_height = CheckCorner(x1, z1, x_plane, z_plane, const_plane);
+        max_y = std::max(max_y, corner_height);
+
+        corner_height = CheckCorner(x1, z2, x_plane, z_plane, const_plane);
+        max_y = std::max(max_y, corner_height);
+
+        corner_height = CheckCorner(x2, z1, x_plane, z_plane, const_plane);
+        max_y = std::max(max_y, corner_height);
+
+        corner_height = CheckCorner(x2, z2, x_plane, z_plane, const_plane);
+        max_y = std::max(max_y, corner_height);
```

Comment:
You can just inline these into the param for max, not much point assigning them to a variable